### PR TITLE
Refactor the proxy options

### PR DESCRIPTION
**Pull request by saragluna**
Proxy authentication type is not supported to set in HTTP-based client's configuration, this PR is to address this issue.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/properties/AzureGlobalProperties.java [hunk 1]
```diff
@@ -145,6 +150,25 @@ public void setNonProxyHosts(String nonProxyHosts) {
         }
     }
 
+    /**
+     * Proxy properties only apply to http-based clients.
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/properties/AzureGlobalProperties.java:154]: http-based? why

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/core/implementation/converter/AzureAmqpProxyOptionsConverter.java [hunk 2]
```diff
@@ -32,18 +33,25 @@ public ProxyOptions convert(ProxyOptionsAware.Proxy proxy) {
...
+                        break;
+                    default:
+                }
+            }
         }
         Proxy.Type type;
+        if (proxy.getType() == null) {
+            throw new IllegalArgumentException("Wrong proxy type provided!");
+        }
         switch (proxy.getType()) {
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/core/implementation/converter/AzureAmqpProxyOptionsConverter.java:55]: do we need to add a "socks5" case here?

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/core/implementation/converter/AzureAmqpProxyOptionsConverter.java [hunk 3]
```diff
@@ -32,18 +33,25 @@ public ProxyOptions convert(ProxyOptionsAware.Proxy proxy) {
...
+                        break;
+                    case "digest":
+                        authenticationType = ProxyAuthenticationType.DIGEST;
+                        break;
+                    default:
+                }
+            }
         }
         Proxy.Type type;
+        if (proxy.getType() == null) {
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/core/implementation/converter/AzureAmqpProxyOptionsConverter.java:52]: do we need to add a "socks5" case here?

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/core/implementation/converter/AzureAmqpProxyOptionsConverter.java:52]: No for now, the underline amqp options doesn't have a socks4 or sock5, they only have one socks

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/converter/AzureAmqpProxyOptionsConverter.java [hunk 4]
```diff
@@ -32,18 +33,25 @@ public ProxyOptions convert(ProxyOptionsAware.Proxy proxy) {
...
+                        break;
+                    case "digest":
+                        authenticationType = ProxyAuthenticationType.DIGEST;
+                        break;
+                    default:
+                }
+            }
         }
         Proxy.Type type;
+        if (proxy.getType() == null) {
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/converter/AzureAmqpProxyOptionsConverter.java:52]: I suggest moving the part before 
```
ProxyAuthenticationType authenticationType = ProxyAuthenticationType.NONE;
...
```

File: sdk/spring/spring-cloud-azure-core/src/test/java/com/azure/spring/core/implementation/converter/AzureHttpProxyOptionsConverterTests.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        source.setPassword("my-password");
+        source.setUsername("my-username");
+        source.setType("socks4");
+
+        ProxyOptions target = AzureHttpProxyOptionsConverter.HTTP_PROXY_CONVERTER.convert(source);
+        assertNull(target);
+    }
+
+    @Test
+    void correctlyConvertProxyType() {
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-core/src/test/java/com/azure/spring/core/implementation/converter/AzureHttpProxyOptionsConverterTests.java:65]: strange for me with the method name "correctlyConvertProxyType"

**Review by backwind1233 — APPROVED** [review]
LGTM

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/converter/AzureAmqpProxyOptionsConverter.java (hunk 4 above)
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/converter/AzureAmqpProxyOptionsConverter.java:52]: Good point, will change in another pr.

File: sdk/spring/spring-cloud-azure-core/src/test/java/com/azure/spring/core/implementation/converter/AzureHttpProxyOptionsConverterTests.java (hunk 5 above)
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-core/src/test/java/com/azure/spring/core/implementation/converter/AzureHttpProxyOptionsConverterTests.java:65]: Any suggestions?

Comment (backwind1233) [reply sdk/spring/spring-cloud-azure-core/src/test/java/com/azure/spring/core/implementation/converter/AzureHttpProxyOptionsConverterTests.java:65]: no, just keep the name.

**Comment by backwind1233** [comment]
I have a simple question? Do we have IT to test the proxy options?

**Comment by saragluna** [comment]
> I have a simple question? Do we have IT to test the proxy options?

No, we don't.

**Comment by saragluna** [comment]
> Why do we need spring-cloud-azure-core dependency in autoconfigure module?

We will depend on it transitively, but the current 4.0 branch has a version conflict on azure-core. The version of azure-core on 4.0 branch is 1.25.0, but the transitive dependey version of it from sdks is 1.24.1, this is why I add it in the first place.

File: sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/converter/AzureAmqpProxyOptionsConverter.java (hunk 4 above)
Comment (saragluna) [reply sdk/spring/spring-cloud-azure-core/src/main/java/com/azure/spring/cloud/core/implementation/converter/AzureAmqpProxyOptionsConverter.java:52]: https://github.com/Azure/azure-sdk-for-java/pull/27429/files

**Pull request merged**